src/backtracking/knight_tour.rs

```diff
@@ -0,0 +1,182 @@
+//! This module contains the implementation of the Knight's Tour problem.
+//!
+//! The Knight's Tour is a classic chess problem where the objective is to move a knight to every square on a chessboard exactly once.
+
+/// Finds the Knight's Tour starting from the specified position.
+///
+/// # Arguments
+///
+/// * `size_x` - The width of the chessboard.
+/// * `size_y` - The height of the chessboard.
+/// * `start_x` - The x-coordinate of the starting position.
+/// * `start_y` - The y-coordinate of the starting position.
+///
+/// # Returns
+///
+/// A tour matrix if the tour was found or None if not found.
+/// The tour matrix returned is essentially the board field of the `KnightTour`
+/// struct `Vec<Vec<i32>>`. It represents the sequence of moves made by the
+/// knight on the chessboard, with each cell containing the order in which the knight visited that square.
+pub fn find_knight_tour(
+    size_x: usize,
+    size_y: usize,
+    start_x: usize,
+    start_y: usize,
+) -> Option<Vec<Vec<usize>>> {
+    let mut tour = KnightTour::new(size_x, size_y);
+    tour.find_tour(start_x, start_y)
+}
+
+/// Represents the KnightTour struct which implements the Knight's Tour problem.
+struct KnightTour {
+    board: Vec<Vec<usize>>,
+}
+
+impl KnightTour {
+    /// Possible moves of the knight on the board
+    const MOVES: [(i32, i32); 8] = [
+        (2, 1),
+        (1, 2),
+        (-1, 2),
+        (-2, 1),
+        (-2, -1),
+        (-1, -2),
+        (1, -2),
+        (2, -1),
+    ];
+
+    /// Constructs a new KnightTour instance with the given board size.
+    /// # Arguments
+    ///
+    /// * `size_x` - The width of the chessboard.
+    /// * `size_y` - The height of the chessboard.
+    ///
+    /// # Returns
+    ///
+    /// A new KnightTour instance.
+    fn new(size_x: usize, size_y: usize) -> Self {
+        let board = vec![vec![0; size_x]; size_y];
+        KnightTour { board }
+    }
+
+    /// Checks if the given position is safe to move to.
+    ///
+    /// # Arguments
+    ///
+    /// * `x` - The x-coordinate of the position.
+    /// * `y` - The y-coordinate of the position.
+    ///
+    /// # Returns
+    ///
+    /// A boolean indicating whether the position is safe to move to.
+    fn is_safe(&self, x: i32, y: i32) -> bool {
+        x >= 0
+            && y >= 0
+            && x < self.board.len() as i32
+            && y < self.board[0].len() as i32
+            && self.board[x as usize][y as usize] == 0
+    }
+
+    /// Recursively solves the Knight's Tour problem.
+    ///
+    /// # Arguments
+    ///
+    /// * `x` - The current x-coordinate of the knight.
+    /// * `y` - The current y-coordinate of the knight.
+    /// * `move_count` - The current move count.
+    ///
+    /// # Returns
+    ///
+    /// A boolean indicating whether a solution was found.
+    fn solve_tour(&mut self, x: i32, y: i32, move_count: u32) -> bool {
```

Comment:
`move_count` should be of the type `usize`. Also `x` and `y` should be of the type [`isize`](https://doc.rust-lang.org/std/primitive.isize.html#).

Reply:
It's a good refactoring, changes are clarity and consistency.